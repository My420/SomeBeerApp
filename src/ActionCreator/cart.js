import {
  CART_ADD_ITEM,
  CART_DELETE_ITEM,
  CART_SET_ITEM_AMOUNT
} from '../utils/constants';

export const addToCart = itemData => {
  return {
    type: CART_ADD_ITEM,
    payload: {
      itemData
    }
  };
};

export const deleteFromCart = itemId => {
  return {
    type: CART_DELETE_ITEM,
    payload: {
      itemId
    }
  };
};

export const changeCartItemAmount = (itemId, amount) => {
  return {
    type: CART_SET_ITEM_AMOUNT,
    payload: {
      itemId,
      amount
    }
  };
};
