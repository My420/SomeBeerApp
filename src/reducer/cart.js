import { Record, Map } from 'immutable';
import {
  CART_ADD_ITEM,
  CART_DELETE_ITEM,
  CART_SET_ITEM_AMOUNT
} from '../utils/constants';

import BeerDataRecord from '../utils/BeerDataRecord';

export const cartInitialState = Map({});

export const cartItemRecord = Record(
  {
    amount: 0,
    item: null
  },
  'cartItemRecord'
);

const cart = (state = cartInitialState, action) => {
  const { type, payload } = action;
  switch (type) {
    case CART_ADD_ITEM: {
      const { itemData } = payload;
      const { id } = itemData;
      const item = cartItemRecord({
        amount: 1,
        item: BeerDataRecord(itemData)
      });

      return state.set(id, item);
    }
    case CART_DELETE_ITEM: {
      const { itemId } = payload;

      return state.delete(itemId);
    }
    case CART_SET_ITEM_AMOUNT: {
      const { itemId, amount } = payload;
      const newItem = state.get(itemId).set('amount', amount);

      return state.set(itemId, newItem);
    }

    default:
      return state;
  }
};

export default cart;
