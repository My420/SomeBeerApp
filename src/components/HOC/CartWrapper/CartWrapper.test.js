import React from 'react';
import { Provider } from 'react-redux';
import { createStore, combineReducers } from 'redux';
import { mount } from 'enzyme';
import withCart from './CartWrapper';
import cart, { cartInitialState } from '../../../reducer/cart';

describe('Test withCart HOC', () => {
  const TestComp = () => {
    return <p>My Component</p>;
  };

  const ComponentWrapper = withCart(TestComp);
  const store = createStore(combineReducers({ cart }), {});

  const wrapper = mount(
    <Provider store={store}>
      <ComponentWrapper test="test" anotherOne="anotherOne" />
    </Provider>
  );

  test('should correct render component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should render correct name for ComponentWrapper component', () => {
    const component = wrapper.find('CartWrapper(TestComp)');
    expect(component).toHaveLength(1);
  });

  test('should render TestComp inside ComponentWrapper component', () => {
    const component = wrapper.find('CartWrapper(TestComp)').find('TestComp');
    expect(component).toHaveLength(1);
  });

  test('ComponentWrapper should pass props through to TestComp component', () => {
    const component = wrapper.find('TestComp');
    expect(component.prop('test')).toBe('test');
    expect(component.prop('anotherOne')).toBe('anotherOne');
  });

  test('TestComp component should have correct cartData prop', () => {
    const component = wrapper.find('TestComp');
    expect(component.prop('cartData')).toEqual(cartInitialState);
  });

  test('addItemToCart prop in TestComp component should be function', () => {
    const component = wrapper.find('TestComp');
    expect(typeof component.prop('addItemToCart')).toBe('function');
  });

  test('deleteItemFromCart prop in TestComp component should be function', () => {
    const component = wrapper.find('TestComp');
    expect(typeof component.prop('deleteItemFromCart')).toBe('function');
  });

  test('changeCartItemAmount prop in TestComp component should be function', () => {
    const component = wrapper.find('TestComp');
    expect(typeof component.prop('changeCartItemAmount')).toBe('function');
  });
});
