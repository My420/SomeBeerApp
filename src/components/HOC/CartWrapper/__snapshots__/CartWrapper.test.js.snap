// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test withCart HOC should correct render component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(CartWrapper(TestComp))
    anotherOne="anotherOne"
    test="test"
  >
    <CartWrapper(TestComp)
      addItemToCart={[Function]}
      anotherOne="anotherOne"
      cartData={Immutable.Map {}}
      changeCartItemAmount={[Function]}
      deleteItemFromCart={[Function]}
      test="test"
    >
      <TestComp
        addItemToCart={[Function]}
        anotherOne="anotherOne"
        cartData={Immutable.Map {}}
        changeCartItemAmount={[Function]}
        deleteItemFromCart={[Function]}
        test="test"
      >
        <p>
          My Component
        </p>
      </TestComp>
    </CartWrapper(TestComp)>
  </Connect(CartWrapper(TestComp))>
</Provider>
`;
